tests/00expect_http_fail.pl

```diff
@@ -114,18 +114,68 @@ sub check_http_code
    );
 }
 
+=head2 expect_matrix_error
 
-# todo: generalise this to other http errors/matrix errcodes
-sub expect_m_not_found
+   http_request()->main::expect_matrix_error(
+      'M_EXPECTED_ERROR',
+      http_code => 418,
+   )->then( sub {
+      my ( $error_body ) = @_;
+   });
+
+A decorator for a Future which we expect to return an HTTP error with a Matrix
+error code. Asserts that the HTTP error code and matrix error code were as
+expected, and if so returns the JSON-decoded body of the error (so that it can
+be checked for additional fields).
+
+By default, the expected HTTP error code is 400. This can be overridden with an
+"http_code" parameter.
+
+=cut
+
+sub expect_matrix_error
 {
-   my $f = shift;
+   my ( $f, $expected_errcode, %opts ) = @_;
+
+   my $expected_http_code = $opts{http_code} // 400;
```

Comment:
I think it would be clearer to always require the HTTP code, so it becomes something like:

```perl
)->expect_matrix_error(418, 'M_EXPECTED_ERROR')
```